common/app/common/inject/HttpContextScope.java

```diff
@@ -0,0 +1,97 @@
+package common.inject;
+
+import com.google.common.collect.Maps;
+import com.google.inject.Key;
+import com.google.inject.Provider;
+import com.google.inject.Scope;
+import com.google.inject.Scopes;
+import play.mvc.Http;
+
+import java.util.Map;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+/*
+
+IMPORTANT: put @ before the route
+
+inspirated by: http://stackoverflow.com/questions/25626264/how-to-use-play-frameworks-request-and-session-scope-in-google-guice/34981902#34981902
+ */
+
+
+/**
+ * Allows objects to be bound to Play! Http.Context.current.args with a ThreadLocal fallback.
+ */
+public class HttpContextScope implements Scope {
+
+    private static final ThreadLocal<Context> httpContextScopeContext = new ThreadLocal<>();
+
+    enum NullableObject {
+        INSTANCE
+    }
+
+    @Override
+    public <T> Provider<T> scope(final Key<T> key, final Provider<T> provider) {
+        return new Provider<T>() {
+            @Override
+            public T get() {
+                Http.Context currentContext = Http.Context.current();
+                if (currentContext == null) {
+                    Context context = httpContextScopeContext.get();
+                    if (context != null) {
+                        T t = (T) context.map.get(key);
```

Comment:
I'm confused, because nothing is warning me against it, which makes me think it wouldn't, but how? I don't see it